### PR TITLE
Add Specific TypeError for Missing Type Classes

### PR DESCRIPTION
**Pull request by mlutze**
fixes https://github.com/flix/flix/issues/1552

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/errors/TypeError.scala [hunk 1]
```diff
@@ -274,4 +274,24 @@ object TypeError {
...
+    def summary: String = s"No instance of class '$clazz' for type '${FormatType.formatType(tpe)}'."
+
+    def message: VirtualTerminal = {
+      val vt = new VirtualTerminal()
+      vt << Line(kind, source.format) << NewLine
+      vt << ">> No instance of class '" << Red(clazz.toString) << "' for type '" << Red(FormatType.formatType(tpe)) << "'." << NewLine
+      vt << NewLine
+      vt << Code(loc, "no instance found") << NewLine
+      vt << NewLine
+      vt << Underline("Tip:") << " Add an instance for the type." << NewLine
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/errors/TypeError.scala:294]: Quick comment: We could list the available instances. Thoughts?

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/errors/TypeError.scala:294]: What happens if we have 100 instances? The flood might be inconvenient.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/errors/TypeError.scala:294]: I guess this is a problem for say `Eq` but less if you define your own type class.

File: main/src/ca/uwaterloo/flix/language/phase/unification/ClassEnvironment.scala [hunk 2]
```diff
@@ -32,19 +32,19 @@ object ClassEnvironment {
...
+      ().toOk
     } else {
       // Case 2: there is an instance matching tconstr and all of the instance's constraints are entailed by tconstrs0
-      byInst(tconstr, instances) match {
-        case Result.Ok(tconstrs) => tconstrs.forall(entail(tconstrs0, _, instances))
-        case Result.Err(_) => false
-      }
+      for {
+        tconstrs <- byInst(tconstr, instances)
+        _ <- Result.sequence(tconstrs.map(entail(tconstrs0, _, instances)))
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/ClassEnvironment.scala:46]: I think this only reports the first (for some definition of first) instance error. Maybe we should use Validation instead?

File: main/src/ca/uwaterloo/flix/language/phase/unification/ClassEnvironment.scala [hunk 3]
```diff
@@ -56,7 +56,7 @@ object ClassEnvironment {
     @tailrec
     def loop(tconstrs0: List[Ast.TypeConstraint], acc: List[Ast.TypeConstraint]): List[Ast.TypeConstraint] = tconstrs0 match {
       case Nil => acc
-      case head :: tail if entail(acc ++ tail, head, instances) => loop(tail, acc)
+      case head :: tail if entail(acc ++ tail, head, instances).isInstanceOf[Result.Ok[_, _]] => loop(tail, acc)
       case head :: tail => loop(tail, head :: acc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/ClassEnvironment.scala:60]:
```
case head :: tail => entail(acc ++ tail, head, instances) match {
  case Ok(_) =>  loop(tail, acc)
  case Err(_) => loop(tail, head :: acc)
}
```

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/unification/ClassEnvironment.scala:60]: And maybe add a comment for each case that explains what is happening.

File: main/src/ca/uwaterloo/flix/language/phase/unification/ClassEnvironment.scala [hunk 4]
```diff
@@ -32,19 +32,19 @@ object ClassEnvironment {
     * That is, `tconstr` is true if all of `tconstrs0` are true.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/ClassEnvironment.scala:32]: Update documentation

File: main/src/ca/uwaterloo/flix/language/ast/Scheme.scala [hunk 5]
```diff
@@ -128,31 +127,29 @@ object Scheme {
     */
   // TODO can optimize?
   def equal(sc1: Scheme, sc2: Scheme, classEnv: Map[Symbol.ClassSym, List[Ast.Instance]])(implicit flix: Flix): Boolean = {
-    lessThanEqual(sc1, sc2, classEnv) && lessThanEqual(sc2, sc1, classEnv)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Scheme.scala:131]: If checkLessEqual returns a validation then these could be seqM'ed to combine the errors.

File: main/src/ca/uwaterloo/flix/language/ast/Scheme.scala [hunk 6]
```diff
@@ -128,31 +127,29 @@ object Scheme {
...
     * Returns `true` if the given scheme `sc1` is smaller or equal to the given scheme `sc2`.
     */
-  def lessThanEqual(sc1: Scheme, sc2: Scheme, classEnv: Map[Symbol.ClassSym, List[Ast.Instance]])(implicit flix: Flix): Boolean = {
-
+  def checkLessThanEqual(sc1: Scheme, sc2: Scheme, classEnv: Map[Symbol.ClassSym, List[Ast.Instance]])(implicit flix: Flix): Result[Unit, UnificationError] = {
 
     /**
       * Checks that `tconstr` is entailed by `tconstrs`.
       */
     def checkEntailment(tconstrs: List[Ast.TypeConstraint], tconstr: Ast.TypeConstraint): Result[Unit, UnificationError] = {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Scheme.scala:144]: I guess there is no need for this function anymore?

File: main/src/ca/uwaterloo/flix/language/errors/TypeError.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/errors/TypeError.scala:294]: I think we leave it as is for now.

**Comment by magnus-madsen** [comment]
Very nice!

**Pull request merged**